src/main/java/com/puppycrawl/tools/checkstyle/site/ParentModuleMacro.java

```diff
@@ -0,0 +1,96 @@
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.util.Locale;
+import java.util.Set;
+
+import com.puppycrawl.tools.checkstyle.ModuleFactory;
+import com.puppycrawl.tools.checkstyle.PackageNamesLoader;
+import com.puppycrawl.tools.checkstyle.PackageObjectFactory;
+import com.puppycrawl.tools.checkstyle.TreeWalkerFilter;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.Filter;
+import org.apache.maven.doxia.macro.AbstractMacro;
+import org.apache.maven.doxia.macro.Macro;
+import org.apache.maven.doxia.macro.MacroExecutionException;
+import org.apache.maven.doxia.macro.MacroRequest;
+import org.apache.maven.doxia.sink.Sink;
+import org.codehaus.plexus.component.annotations.Component;
+
+/**
+ * A macro that inserts a link to the parent module.
+ */
+@Component(role = Macro.class, hint = "parent-module")
+public class ParentModuleMacro extends AbstractMacro {
+    @Override
+    public void execute(Sink sink, MacroRequest request) throws MacroExecutionException {
+        final String moduleName = (String) request.getParameter("moduleName");
+        final ModuleFactory moduleFactory;
+        try {
+            final ClassLoader cl = ParentModuleMacro.class.getClassLoader();
+            final Set<String> packageNames = PackageNamesLoader.getPackageNames(cl);
+            moduleFactory = new PackageObjectFactory(packageNames, cl);
+            final Object instance = moduleFactory.createModule(moduleName);
+            final Class<?> clss = instance.getClass();
+            System.out.println(clss.getSuperclass());
+            sink.paragraph();
+            sink.link("");
+            sink.text(getParentModule(clss, moduleName));
+            sink.link_();
+            sink.paragraph_();
+        }
+        catch (CheckstyleException ex) {
+            throw new MacroExecutionException("Failed to create module factory", ex);
+        }
+    }
+
+    /**
+     * Returns the parent module name for the given module class. Returns either
+     * "TreeWalker" or "Checker". Returns null if the module class is null.
+     *
+     * @param moduleClass the module class.
+     * @param moduleName the module name we are looking for the parent of.
+     * @return the parent module name as a string.
+     * @throws MacroExecutionException if the parent module cannot be found.
+     */
+    // -@cs[ReturnCount] Too complex to break apart.
+    private static String getParentModule(Class<?> moduleClass, String moduleName)
```

Comment:
If it is this painful to determine parent module, we should consider some way to attach some metadata (via annotation or otherwise) to each check.

Reply:
I wouldn't call it painful. The `ReturnCount` check is not doing any help here in my opinion. I should change the suppression message to something else. This one I copied from somewhere in the codebase.

Reply:
I would consider having to check for this many different classes and having a tail recursive call to this method painful :) It is a matter of opinion for sure, but it seems like it should be easier to figure this out.

Reply:
I would prefer some iterative way to do this, like keep getting the super class until we hit X or Y, then return the value based on that.

Reply:
Done.